### PR TITLE
graphSage demo setup

### PR DESCRIPTION
**Pull request by ff-l**
### Current implementation
- [x] Data extract (28-31/01/19): graphSAGE was trained on a filtered subset (to ensure pages exist/are content items)
- [x] Analyse produced functional network 
- [x] Data extraction module for Content API (to get page text, links and metadata out)
- [x] Train graphSAGE on functional network
### Post-MVP/extra training
- [ ] Build structural network from extracted page links
- [ ] Compute structural network coverage
- [ ] Make script

### CODE REVIEW

File: src/data/content_api_extract.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        pd.concat([pd.read_csv(f, compression="gzip") for f in filelist], ignore_index=True).to_csv(filename,
+                                                                                                    compression="gzip",
+                                                                                                    index=False)
+        all_files.extend(filelist)
+    return all_files
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser(description='Module that produces a merged, metadata-aggregated and '
```
Comment (mammykins) [inline src/data/content_api_extract.py]: description needs changing

Comment (ff-l) [reply src/data/content_api_extract.py]: 👍

**Review by mammykins — COMMENTED** [review]
This PR uses a page list provided by a node_list (where node_id is it's url), presumably derived from the functional network GOV.UK (network-data). Using this list the PR then uses the content api to look up and return a nested data structure with the python package beautiful soup. This gives access to all the text content on a page and links etc.

With this data structure associated with each node, the PR then works out what in-text and related links are present on the page (the structural network?).

For a better understanding of the metadata produced by the content api one can use the analyse_note_metadata notebook to review. The data includes url, contentid (a unique identifier). It focuses on code to count unique doc types and also how to extract test data associated with a page, such as the title.  The comprehensive use of code to explore the data helps visualise this, for example using `.head()` of the dataframes. These act as informal tests of code quality and that it is behaving as expected.

Would using the contentid be a better node-id than url? However, I expect it might be difficult to join urls derived from BigQuery data with the contentapi derived contentid and urls; they're might be a mismatch. This is demonstrated in your training of the model where you drop missing pages after joining urls in this way. 

The notebook `analyse_funct_network` looks to tidy our nodes in the functional network. It drops nodes that contain spaces " " and search terms, it does look to keep "/search?q=". Is this a useful node to keep? Should it be included in the training, or is the intention to drop it from the list of predicted links? (related links shouldn't point users to search)

The codebase looks to reuse some work from our previous networkx notebooks in the data pipeline repo. Attributes are added to nodes as standard.

Some network statistics generation code is present, but presumably these take a while to run... Interesting stats on degree distribution etc. This knowledge is handy to have for the GOV.UK functional network. We should look to running the average shortest path and Djkistra's at some point.

The graph sage implementation takes an edge list of the functional network (derives node list from this also) and adds attributes to each node in the graph. The attributes are the term frequency–inverse document frequency (the tfidf vectorise from sklearn is used on each nodes title and description text from the contentapi, after removing stopwords; taking the top 2000 words as input, a one-hot encoded 2000 length vector?). This completes the graph and provides it with some simple attributes to begin training and testing graphsage.

The training and testing sets are created using stellargraph package functions. The notebooks for training and testing are based on community available demo notebooks from stellargraph. This provides a structure and flow to the work with some useful markdown comments explaining the "why". However some of these comments aren't relevant to our example and have some erroneous comments. For example, when describing the number of hidden layers and dropout there is a mismatch between the markdown and the code. This should be corrected.

The model parameters are set. The PR does not use defaults from the demo examples but instead looks to reduce the number of parameters of the model, presumably in an attempt to reduce the computation time for training?

After the parameters were set, data generators were created. The activation function for the output was sigmoid, to give probability that an edge does exist? The GraphSAGE model acts like another layer in a deep learning model (I guess as a node embedding it has a similar role to a word embedding layer). It is added into the Keras stack of layers.

The prediction notebook is harder to follow due to lack of markdown or comments as to the why. However, it follows on from the demos provided by stellargraph so I don't think this is too much of an issue.

I need to review these demos then revisit the training and testing aspects to offer further insight. I suggest you merge for now after tackling some of the issues above. The documentation and commenting in the training and testing notebooks could be added as an issue and tackled with a later PR after an mvp of the graphsage has been used to predict links.

**Comment by ff-l** [comment]
@mammykins Amazing, responses/actions:

- **Would using the contentid be a better node-id than url?** I think this need to be dealt with soon. Apart from visualization/our own internal understanding, depending on page urls for identification is not a great idea. Will look into how this direct extract can happen (there is some uncertainty as to whether the custom dimension with the ids has been validated/is reliable)
- **it does look to keep "/search?q=".** all search pages are dropped from the journeys in the govuk-network-data `make_network_data.py` script.
- **the tfidf vectorise from sklearn is used on each nodes title and description text from the contentapi, after removing stopwords; taking the top 2000 words as input, a one-hot encoded 2000 length vector?** sort of, apart from the last bit, not one-hot encoded, tf-idf vector containing the tf-idfs (ha, I mean weights) for each of the top 2K most frequent words. These values are also normalized. Any word that appears in more than 80% of samples is also dropped. Not great, would be better to use a more robust approach in future, like doc2vec or embeddings (embeddings within embeddings 🚀)
- **The model parameters are set. The PR does not use defaults from the demo examples but instead looks to reduce the number of parameters of the model, presumably in an attempt to reduce the computation time for training?** not the case, which bit are you talking about?
- **The activation function for the output was sigmoid, to give probability that an edge does exist?** sigmoid of the inner dot product of the node vectors representing candidate edge. Will experiment with other vector combo methods (dot is pretty standard etc)
- **The prediction notebook is harder to follow due to lack of markdown or comments as to the why. However, it follows on from the demos provided by stellargraph so I don't think this is too much of an issue.** apols, the other #10 has a much better implementation + script

**Comment by mammykins** [comment]
thanks, I'll update the google doc with this info
